lua/visual.lua

```diff
@@ -153,7 +153,19 @@ core.register_entity("drawers:visual", {
 	end,
 
 	on_rightclick = function(self, clicker)
-		if core.is_protected(self.drawer_pos, clicker:get_player_name()) then
+		local upgrades = core.get_meta(self.drawer_pos):get_inventory():get_list("upgrades")
+
+		local private = true
+		for _,itemStack in pairs(upgrades) do
+			local iname = itemStack:get_name()
+			local idef = core.registered_items[iname]
+			if idef.groups.drawer_public == 1 then
```

Comment:
`idef` might be `nil` when items are removed from the game. This is an edge-case, but yet a potential bug.

Also I suggest to move all protection checks to a local function so that `on_rightclick` and `on_punch` use the identical checks.